File: 08-04-2024/Optimal Strategy For A Game.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    for(let i=0;i<t;i++)
    {
        let input_line = readLine().split(' ');
        let n = parseInt(input_line[0]);
        
        input_line = readLine().split(' ');
        let arr = new Array(n);
        for(let i=0;i<n;i++)
            arr[i] = parseInt(input_line[i]);
        
        
        let obj = new Solution();
        let ans = obj.maximumAmount(arr, n);
        if(ans==-0n)
            ans=0n;
        console.log(ans.toString());
    }
}
// } Driver Code Ends




//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @returns {BigInt}
*/

class Solution 
{
    //Function to find the maximum possible amount of money we can win.
    maximumAmount(arr, n)
    { 
        //Your Code Start
          let dp = Array.from({ length: n }, () => Array(n).fill(0));

        for (let g = 0; g < n; g++) {
            for (let i = 0, j = g; j < n; i++, j++) {
                if (g === 0)
                    dp[i][j] = arr[i];
                else if (g === 1)
                    dp[i][j] = Math.max(arr[i], arr[j]);
                else {
                    dp[i][j] = Math.max(
                        arr[i] + Math.min(dp[i + 2][j] || 0, dp[i + 1][j - 1] || 0),
                        arr[j] + Math.min(dp[i + 1][j - 1] || 0, dp[i][j - 2] || 0)
                    );
                }
            }
        }

        return dp[0][n - 1];
         // Your code End
    
    
      
    }
}

File: 08-04-2024/Optimal Strategy For A Game.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

//Function to find the maximum possible amount of money we can win.
class Solution{
    public:
    long long maximumAmount(int n, int arr[]){
        // Your code Start
        int dp[n][n];

        for (int g = 0; g < n; g++) {
            for (int i = 0, j = g; j < n; i++, j++) {
                if (g == 0)
                    dp[i][j] = arr[i];
                else if (g == 1)
                    dp[i][j] = max(arr[i], arr[j]);
                else {
                    dp[i][j] = max(
                        arr[i] + min(dp[i + 2][j], dp[i + 1][j - 1]),
                        arr[j] + min(dp[i + 1][j - 1], dp[i][j - 2])
                    );
                }
            }
        }

        return dp[0][n - 1];
    //Your Code End
        
    }
};


//{ Driver Code Starts.
int main() 
{
   
   	int t;
	cin>>t;
	while(t--)
	{
		int n;
		cin>>n;
		int a[n];
		for(int i=0;i<n;i++)
		cin>>a[i];
		Solution ob;
		cout<< ob.maximumAmount(n,a)<<endl;
	}
	return 0;
}
// } Driver Code Ends
